### ISSUE TITLE
Does not work on Visual Studio 2017 prerelease builds

### ISSUE THREAD

**Issue by spaun2002** [Caphyon/clang-power-tools#160]
Unfortunatelly cannot provide more information on this, as there is nothing else related in the logs.
```
Did not expect that
At C:\Users\***\AppData\Local\Microsoft\VisualStudio\15.0_45d4eb56\Extensions\ieuiy4ub.t1d\clang-build.ps1:1134 
char:9
+         throw "Did not expect that"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Did not expect that:String) [], RuntimeException
    + FullyQualifiedErrorId : Did not expect that
```
[Feature request] It would be really helpful to see more details about exceptions like this. Project or props file that is being parsed, line number (or part of XML) and call stack would be really helpful to understand failures.

**Comment by cpp-red-lion** [comment]
VS Settings:
`[TOOLS] > [Options] > [Clang Power Tools] > [General] > Verbose Mode = True`

**Comment by cpp-red-lion** [comment]
What version of Clang Power Tools are you using ?

Please attach your `.vcxproj` file and (if the case) the required VS property-sheets, so we can debug the issue.

**Comment by spaun2002** [comment]
Verbose didn't add a lot of info (btw, why it resets to default value after update?):
```
VERBOSE: SET_VAR ProjectGuid: {1D8E156F-6EDC-40EA-BDBC-6B0175542760}
VERBOSE: SET_VAR RootNamespace: ###
VERBOSE: SET_VAR ReferencePath: 
Z:\src\***\;Z:\src\***\;Z:\src\***;
VERBOSE: SET_VAR CharacterSet: MultiByte
VERBOSE: SET_VAR ConfigurationType: StaticLibrary
VERBOSE: SET_VAR MSBuildThisFileFullPath: Z:\src\***\###.vcxproj
VERBOSE: SET_VAR MSBuildThisFileExtension: .vcxproj
VERBOSE: SET_VAR MSBuildThisFile: ###.vcxproj
VERBOSE: SET_VAR MSBuildThisFileName: ###
VERBOSE: SET_VAR MSBuildThisFileDirectory: Z:\src\***
VERBOSE: Force includes: stdafx.h bond\core\warning.h
VERBOSE: PCH cpp name: stdafx.cpp
VERBOSE: PCH header name: stdafx.h
VERBOSE: PCH directory: Z:\src\***\
Did not expect that
At C:\Users\*****\AppData\Local\Microsoft\VisualStudio\15.0_45d4eb56\Extensions\ieuiy4ub.t1d\clang-build.ps1:1134 
char:9
+         throw "Did not expect that"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Did not expect that:String) [], RuntimeException
    + FullyQualifiedErrorId : Did not expect that
```

I really doubt I can attach project file due to security reasons and project includes dosens of props files with dosens of dependencies each (so hundreds of props)

**Comment by spaun2002** [comment]
> What version of Clang Power Tools are you using ?
The latest available at this moment: 2.2.0

**Comment by cpp-red-lion** [comment]
> I really doubt I can attach project file due to security reasons and project includes dosens of props files with dosens of dependencies each (so hundreds of props)

I don't think we could easily pinpoint the problem in your project without the actual `vcxproj` config files.
We can try, but we need to exchange a lot more VERBOSE logs :(

**Comment by hero101111** [comment]
Hi,

The MSBuild interpreter mechanism is able to put out an even more verbose log, called _debug log_.
Right now, we're not exposing this option in the VS Extension UI, as it is extremely verbose.

To activate debug mode you have to manually edit the `clang-build.ps1` script inside the VS extension directory. To locate the extension directory search for `clang-build.ps1` in `%LOCALAPPDATA%`

Open the script in an editor and add the following line, after the script `Param()` block
`$DebugPreference = "Continue"`

Restart VS and compile again. You should see a more detailed log.

**Comment by cpp-red-lion** [comment]
Hi,

@spaun2002, have you tried to run with _Debug Log_ mode on, as @hero101111 instructed you above ?
We don't seem to make any progress on this issue :(

I'm afraid I'm going to have to close it, if we don't have sufficient info to reproduce.

**Comment by spaun2002** [comment]
Those are the only meaningful error I could found in the verbose logs:
```
Get-Item : Cannot find path 'Z:\VC\Tools\MSVC\' because it does not exist.
At C:\Users\user_name\AppData\Local\Microsoft\VisualStudio\15.0_45d4eb56\Extensions\asdvb0ky.1hd\clang-build.ps1:582 
char:12
+   return ((Get-Item "$(Get-VisualStudio-Path)\VC\Tools\MSVC\" | Get-C ...
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Z:\VC\Tools\MSVC\:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand
    
    
VERBOSE: MSCVER: 
Get-VisualStudio-Includes : Cannot bind argument to parameter 'vsPath' because it is an empty string.
At C:\Users\user_name\AppData\Local\Microsoft\VisualStudio\15.0_45d4eb56\Extensions\asdvb0ky.1hd\clang-build.ps1:852 
char:55
+     $returnArray += Get-VisualStudio-Includes -vsPath $vsPath -mscVer ...
+                                                       ~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-VisualStudio-Includes], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Get-VisualStudio-Includes
    
VERBOSE: Generating PCH...
VERBOSE: SET_VAR ClangPowerTools:CppStd: c++14
VERBOSE: PCH creation args: "z:\src\***\src\ProjectName\stdafx.h" -emit-pch -o 
"z:\src\***\src\ProjectName.clang.pch" -std=c++14 -Wall -fms-compatibility-version=19.10
 -fms-compatibility -Wmicrosoft -Wno-invalid-token-paste -Wno-unknown-pragmas -Wno-unused-variable -Wno-unused-value 
-Wno-unused-command-line-argument -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -DWIN64 -D_DEBUG 
-DUSING_APSDK_NAMESPACE -DDEBUG -DWINVER=0x0601 -D_WIN32_WINNT=0x0601 -DWINNT=1 -DWIN32=100 -DWIN32_LEAN_AND_MEAN 
-DNOMINMAX -DGSL_THROW_ON_CONTRACT_VIOLATION -DJENNYA_ENABLE_CBOND_TYPES -DJENNYA_ENABLE_CBOND 
-DBOND_COMPACT_BINARY_PROTOCOL -DBOND_JSON_PROTOCOL -DWIN32_LEAN_AND_MEAN -DNOMINMAX -DBOOST_USE_WINDOWS_H -D_MBCS
clang++.exe: warning: treating 'c-header' input as 'c++-header' when in C++ mode, this behavior is deprecated [-Wdeprecated]
z:\src\***\src\ProjectName\stdafx.h:3:10: fatal error: 'boost/format.hpp' file not found
#include <boost/format.hpp>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
VERBOSE: Cleaning up PCH temporaries:
VERBOSE:   z:\src\***\src\ProjectName.clang.pch
Errors encountered during PCH creation

Done Clang Tidy
```

It seems that Get-VisualStudio-Path fails to find my visual studio instance. I'm using VS2017 Enterprise Preview. VsWhere is not awailable via the %PATH% environment but if I run it with the same arguments as clang-tidy.ps1 does here is what I see:
```
vswhere.exe -nologo -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise
```

**Comment by spaun2002** [comment]
I do not have system wide installed vswhere, so mine is brought by nuget to the following directory: E:\src\CxCache\vswhere.1.0.58
 aVisualStudioVersion = 2017 
 aVisualStudioSku = Enterprise 

Why do you relay on the vswhere package that could be missed on the machine (it is not a part of standard VS installation). You'd better bring it either alongside or download using nuget.exe to somewhere script can use. 
BTW, latest vswhere builds (since 2.0) do not find preview versions without -prerelease argument

**Comment by spaun2002** [comment]
There is C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe file on my machine, but that path is not in the PATH env. variable.

**Comment by cpp-red-lion** [comment]
Thanks for the clarifications, @spaun2002 
We were not aware of `-prerelease` switch for **vswhere**.

We'll fix this ASAP.

**Comment by hero101111** [comment]
Hello,

We've just pushed a fix enabling detection of prerelease versions of Visual Studio 2017.
It will be included in the next release.

Regards,
Gabriel

---

Please note that this fix is currently live only for the github repository.
The latest Visual Studio extension (`2.5.1`) does not contain the fix. It will be included in the next release.

Until then, as a workaround, you can manually replace `clang-build.ps1` in the VS extension directory with the latest from github. To locate the extension directory search for `clang-build.ps1` in `%LOCALAPPDATA%`

Don't forget to restart Visual Studio after replacing the script file.

**Comment by cpp-red-lion** [comment]
Hi,

This issue has been fixed in version `2.6` just released.

Thank you for your feedback 👍
